File: src/Contexts/Basket/Language/Basket/Models/Basket.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace eShop.Basket.Basket.Models
{
    public class Basket
    {
        public Guid Id { get; set; }

        public string CustomerId { get; set; }
        public string Customer { get; set; }

        public int TotalItems => Items?.Count() ?? 0;
        public long TotalQuantity => Items?.Sum(x => x.Quantity) ?? 0L;

        public long SubTotal => Items?.Sum(x => x.SubTotal) ?? 0L;

        public long Created { get; set; }
        public long Updated { get; set; }

        public Entities.Item.Models.BasketItem[] Items { get; set; }
    }
}

File: src/Web/src/app/parts/basket/models/basket.ts
import Debug from 'debug';
import { DateTime } from 'luxon';
import { FormatDefinition } from '../../../components/models';
import { BasketModel as BasketModelBase, BasketType as BasketTypeBase } from '../../../models/basket/baskets';
import { ItemIndexType } from '../../../models/basket/items';

const debug = new Debug('basket');

export interface BasketType extends BasketTypeBase {
  subtractItem: (item: ItemIndexType) => void;
  additionItem: (item: ItemIndexType) => void;

  readonly createdDate?: DateTime;
  readonly updatedDate?: DateTime;
  readonly formatting?: { [idx: string]: FormatDefinition };
}
export const BasketModel = BasketModelBase
  .actions(self => ({
    subtractItem(item: ItemIndexType) {
      self.totalItems--;
      self.totalQuantity -= item.quantity;
      self.subTotal -= item.subTotal;
    },
    additionItem(item: ItemIndexType) {
      self.totalItems++;
      self.totalQuantity += item.quantity;
      self.subTotal += item.subTotal;
    }
  }))
  .views(self => ({
    get createdDate() {
      return DateTime.fromMillis(self.created);
    },
    get updatedDate() {
      return DateTime.fromMillis(self.updated);
    },
    get formatting() {
      return ({
        subTotal: {
          currency: true,
          normalize: 2
        },
        totalFees: {
          currency: true,
          normalize: 2
        },
        totalTaxes: {
          currency: true,
          normalize: 2
        },
        total: {
          currency: true,
          normalize: 2
        }
      });
    },
  }));
